pkg/components/pubsub/pluggable.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	_, err = p.Client.Init(p.Context, &proto.PubSubInitRequest{
+		Metadata: protoMetadata,
+	})
+
+	return err
+}
+
+// Features lists all implemented features.
+func (p *grpcPubSub) Features() []pubsub.Feature {
+	return p.features
+}
+
+// Publish publishes data to a topic.
+func (p *grpcPubSub) Publish(req *pubsub.PublishRequest) error {
+	_, err := p.Client.Publish(p.Context, &proto.PublishRequest{
+		Topic:      req.Topic,
+		PubsubName: req.PubsubName,
+		Data:       req.Data,
+		Metadata:   req.Metadata,
+	})
+	return err
+}
+
+type messageHandler = func(*proto.PullMessageResponse)
+
+// adaptHandler returns a non-error function that handle the message with the given handler and ack when returns.
+//
+//nolint:nosnakecase
+func (p *grpcPubSub) adaptHandler(ctx context.Context, streamingPull proto.PubSub_PullMessagesClient, handler pubsub.Handler) messageHandler {
+	sendLock := &sync.Mutex{}
+	return func(msg *proto.PullMessageResponse) {
+		m := pubsub.NewMessage{
+			Data:        msg.Data,
+			ContentType: &msg.ContentType,
+			Topic:       msg.Topic,
+			Metadata:    msg.Metadata,
+		}
+		var ackError *proto.AckMessageError
+
+		if err := handler(ctx, &m); err != nil {
+			p.logger.Errorf("error when handling message on topic %s", msg.Topic)
+			ackError = &proto.AckMessageError{
+				Message: err.Error(),
+			}
+		}
+
+		// As per documentation:
+		// When using streams,
+		// one must take care to avoid calling either SendMsg or RecvMsg multiple times against the same Stream from different goroutines.
+		// In other words, it's safe to have a goroutine calling SendMsg and another goroutine calling RecvMsg on the same stream at the same time.
+		// But it is not safe to call SendMsg on the same stream in different goroutines, or to call RecvMsg on the same stream in different goroutines.
+		// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams
+		sendLock.Lock()
+		if err := streamingPull.Send(&proto.PullMessagesRequest{
+			AckMessageId: msg.Id,
+			AckError:     ackError,
+		}); err != nil {
+			p.logger.Errorf("error when ack'ing message %s from topic %s", msg.Id, msg.Topic)
+		}
+		sendLock.Unlock()
+	}
+}
+
+// pullMessages pull messages of the given subscription and execute the handler for that messages.
+func (p *grpcPubSub) pullMessages(ctx context.Context, subscription *proto.Subscription, handler pubsub.Handler) error {
+	// first pull should be sync and subsequent connections can be made in background if necessary
+	pull, err := p.Client.PullMessages(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "unable to subscribe")
+	}
+
+	streamCtx, cancel := context.WithCancel(pull.Context())
+
+	err = pull.Send(&proto.PullMessagesRequest{
+		Subscription: subscription,
+	})
+
+	cleanup := func() {
+		if closeErr := pull.CloseSend(); closeErr != nil {
+			p.logger.Warnf("could not close pull stream of topic %s: %v", subscription.Topic, closeErr)
+		}
+		cancel()
```

Comment:
Should we call cancel() even when the first call fails?

Reply:
Yes, especially because it's a good practice to defer cancel right after the creation:

```golang
ctx, cancel := context.WithCancel(ctx)
defer cancel()
```

Cancel also apply some cleanup logic (i.e  removing self reference on parent) in the parent context (received from the runtime) which will be propagated to the entire ctx chain